src/trace/java/app/com/linkedin/playparseq/trace/j/ParSeqTraceBuilderImpl.java

```diff
@@ -11,45 +11,68 @@
 import com.linkedin.playparseq.trace.j.renderers.ParSeqTraceRenderer;
 import com.linkedin.playparseq.trace.j.sensors.ParSeqTraceSensor;
-import com.linkedin.playparseq.trace.utils.ParSeqTraceHelper;
+import com.linkedin.playparseq.trace.utils.PlayParSeqTraceHelper;
 import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
 import java.util.stream.Collectors;
+import javax.inject.Inject;
 import javax.inject.Singleton;
-import play.libs.F;
 import play.mvc.Http;
 import play.mvc.Result;
 
 
 /**
  * The class ParSeqTraceBuilderImpl is an implementation of {@link ParSeqTraceBuilder} with the help from the class
- * {@link ParSeqTraceHelper}.
+ * {@link PlayParSeqTraceHelper}.
  *
  * @author Yinan Ding (yding@linkedin.com)
  */
 @Singleton
-public class ParSeqTraceBuilderImpl extends ParSeqTraceHelper implements ParSeqTraceBuilder {
+public class ParSeqTraceBuilderImpl extends PlayParSeqTraceHelper implements ParSeqTraceBuilder {
 
+  /**
+   * The field _materializer is an Akka Materializer for consuming result body stream.
+   */
+  private final Materializer _materializer;
+
+  /**
+   * The constructor injects the Materializer.
+   *
+   * @param materializer The injected Materializer component
+   */
+  @Inject
+  public ParSeqTraceBuilderImpl(final Materializer materializer) {
+    _materializer = materializer;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
   @SuppressWarnings("unchecked")
   @Override
-  public F.Promise<Result> build(final Http.Context context, final F.Promise<Result> origin,
-      final ParSeqTaskStore parSeqTaskStore, final ParSeqTraceSensor parSeqTraceSensor,
-      final ParSeqTraceRenderer parSeqTraceRenderer) {
+  public CompletionStage<Result> build(final Http.Context context, final CompletionStage<Result> origin,
+                                       final ParSeqTaskStore parSeqTaskStore, final ParSeqTraceSensor parSeqTraceSensor,
+                                       final ParSeqTraceRenderer parSeqTraceRenderer) {
     // Sense
     if (parSeqTraceSensor.isEnabled(context, parSeqTaskStore)) {
       // Bind independent Tasks
-      Set<F.Promise<Object>> promises =
-          parSeqTaskStore.get(context).stream().map(t -> F.Promise.wrap(bindTaskToFuture((Task<Object>) t)))
-              .collect(Collectors.toSet());
+      Set<CompletionStage<Object>> completionStages = parSeqTaskStore.get(context).stream()
+          .map(t -> bindTaskToCompletionStage((Task<Object>) t)).collect(Collectors.toSet());
       // Consume the origin
-      promises.add(origin.flatMap(r -> F.Promise.wrap(consumeResult(r.toScala()))));
+      completionStages.add(origin.thenComposeAsync(r -> consumeResult(r, _materializer)));
       // Render
-      return F.Promise.sequence(promises).flatMap(list -> parSeqTraceRenderer.render(context, parSeqTaskStore));
+      return completionStages.stream().reduce((a, b) -> a.thenCombineAsync(b, (c, d) -> c))
```

Comment:
What I'm doing here is to chain all the CompletionStage together and return a CompletionStage which will be completed when all the CompletionStage are completed. I can add a comment to it. But I don't know if there are better names, because the whole process for a/b/c/d is not very meaningful.

Reply:
Found a better way to achieve this.

Reply:
Great! Much easier to understand now :)